lib/maze/generator.rb

```diff
@@ -0,0 +1,61 @@
+# A thread-safe generator of values, backed by an Enumerator.
+class Generator
+  def initialize(enumerator)
+    # A SizedQueue allows a set number of values to always be ready for clients
+    # and will be automatically topped up from the enumerator.
+    @queue = SizedQueue.new(10)
+
+    # The queue filler continually adds to the queue (when there is room), taking
+    # values from the Enumerator.  This ensure the enumerator is always run inside
+    # the same thread
+    @queue_filler = create_queue_filler(enumerator)
+
+    # a mutex to guard against multiple threads replacing the enumerator at once
+    @semaphore = Thread::Mutex.new
+  end
+
+  # Get the next value
+  #
+  # @return The next value
+  def next
+    @queue.pop
+  end
+
+  # # Set the enumerator used to generate values
+  # #
+  # # @param new_enumerator [Enumerator]
+  # # @return [void]
+  # def enumerator=(new_enumerator)
+  #   @semaphore.synchronize do
+  #     # Stop the current queue filler thread
+  #     @queue_filler.exit
+  #
+  #     # Create a new queue
+  #     @queue.close
+  #     @queue = SizedQueue.new(10) # replace the queue so we get a new empty one
+  #
+  #     # Create a new queue filler with the new enumerator
+  #     @queue_filler = create_queue_filler(new_enumerator)
+  #   end
+  # end
+
+  private
+
+  # Create a thread that will constantly append to @queue with values from the
+  # given enumerator
+  #
+  # @param enumerator [Enumerator]
+  # @return [Thread]
+  def create_queue_filler(enumerator)
+    # By passing the enumerator as an argument to Thread.new, it creates a copy
+    # local to that thread and therefore we're not sharing an enumerator across
+    # threads
+    Thread.new(enumerator) do |status_code_generator|
```

Comment:
This thread is never explicitly killed so it might stay around forever

I'm not actually sure how Ruby handles this — the Generator instance could be garbage collected when Server doesn't have a reference to it anymore, which may kill this thread at the same time, but this thread could also potentially prevent the Generator from being garbage collected at all (especially since it has a reference to `@queue`)

Reply:
I think it's safer to replace the enumerator instance on the Generator, so that it can handle killing the thread internally

Replacing the Generator instance in Server was a neat idea, but if we need to kill this thread explicitly then I think you'll struggle to handle that solely in Server as you'd need to track when you drop the last reference to an existing Generator

Reply:
I've added a `close` method to the Generator now to handle this, which we can call before replacing it.  It means we can keep the internal implementation details of the Generator hidden away (but more importantly I avoid a headache refactoring what I already had).